integration/system/disk_usage_test.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/docker/docker/integration/internal/container"
+	"gotest.tools/v3/assert"
+)
+
+func TestDiskUsage(t *testing.T) {
+	defer setupTest(t)()
+
+	ctx := context.Background()
+	client := testEnv.APIClient()
+
+	baseDU, err := client.DiskUsage(ctx, types.DiskUsageOptions{})
+	assert.NilError(t, err)
+	assert.Assert(t, baseDU.LayersSize > 0)
+	assert.Assert(t, len(baseDU.Images) > 0)
+	assert.DeepEqual(t, baseDU, types.DiskUsage{
+		LayersSize: baseDU.LayersSize,
+		Images:     baseDU.Images,
+		Containers: []*types.Container{},
+		Volumes:    []*types.Volume{},
+		BuildCache: []*types.BuildCache{},
+	})
+
+	type TestCase struct {
+		doc     string
+		options types.DiskUsageOptions
+
+		expected types.DiskUsage
+	}
+	runTests := func(t *testing.T, tcs ...TestCase) {
+		t.Helper()
+
+		for _, tc := range tcs {
+			tc := tc
+			t.Run(tc.doc, func(t *testing.T) {
+				// TODO: Run in parallel once https://github.com/moby/moby/pull/42560 is merged.
+
+				du, err := client.DiskUsage(ctx, tc.options)
+				assert.NilError(t, err)
+				assert.DeepEqual(t, du, tc.expected)
+			})
+		}
+	}
+
+	makeDocFunc := func(prefix string) func(string) string {
+		return func(suffix string) string {
+			return fmt.Sprintf("%s/%s", prefix, suffix)
+		}
+	}
+
+	doc := makeDocFunc("baseline")
+	runTests(t,
+		TestCase{
+			doc: doc("all types"),
+			options: types.DiskUsageOptions{
+				Types: []types.DiskUsageObject{
+					types.BuildCacheObject,
+					types.ContainerObject,
+					types.ImageObject,
+					types.VolumeObject,
+				},
+			},
+			expected: baseDU,
+		},
+
+		TestCase{
+			doc: doc("image types"),
+			options: types.DiskUsageOptions{
+				Types: []types.DiskUsageObject{
+					types.ImageObject,
+				},
+			},
+			expected: types.DiskUsage{
+				LayersSize: baseDU.LayersSize,
+				Images:     baseDU.Images,
+			},
+		},
+
+		TestCase{
+			doc: doc("build-cache types"),
+			options: types.DiskUsageOptions{
+				Types: []types.DiskUsageObject{
+					types.BuildCacheObject,
+				},
+			},
+			expected: types.DiskUsage{
+				BuildCache: []*types.BuildCache{},
+			},
+		},
+
+		TestCase{
+			doc: doc("container, volume types"),
+			options: types.DiskUsageOptions{
+				Types: []types.DiskUsageObject{
+					types.ContainerObject,
+					types.VolumeObject,
+				},
+			},
+			expected: types.DiskUsage{
+				Containers: []*types.Container{},
+				Volumes:    []*types.Volume{},
+			},
+		},
+	)
+
+	cID := container.Run(ctx, t, client)
+
+	du, err := client.DiskUsage(ctx, types.DiskUsageOptions{})
+	assert.NilError(t, err)
+	assert.Assert(t, len(du.Containers) == 1 && du.Containers[0].ID == cID)
+	assert.DeepEqual(t, du, types.DiskUsage{
+		LayersSize: baseDU.LayersSize,
+		Images:     du.Images, // Container counter for one of the images should have increased.
+		Containers: du.Containers,
+		Volumes:    []*types.Volume{},
+		BuildCache: []*types.BuildCache{},
+	})
+
+	doc = makeDocFunc("after container.Run")
+	runTests(t,
+		TestCase{
```

Comment:
These test-cases are all the same, with the exception of the container counter? If so, should se just make a

```
expected := foo
for set := range []string{"before", "after"} {
    for tests {
        t.Run(set + test.doc ())
    }
    run container
    update expected
}
```